lang/LangPrimSource/SC_AudioDevicePrim.cpp

```diff
@@ -125,6 +131,56 @@ int listDevices(struct VMGlobals* g, int type) {
     return 1;
 }
 
+#elif defined(SC_AUDIO_API_PORTAUDIO)
+int listDevices(struct VMGlobals* g, int type) {
+    int numDevices;
+    PaError err = Pa_Initialize();
+    if (err != paNoError)
+        return 0;
+
+    numDevices = Pa_GetDeviceCount();
+    if (numDevices < 0)
+        return 0;
+
+    std::vector<const PaDeviceInfo*> deviceInfos;
+    for (PaDeviceIndex i = 0; i < numDevices; i++) {
+        auto* pdi = Pa_GetDeviceInfo(i);
+        if (type < BOTH) {
+            if (type < IN) {
+                if (pdi->maxOutputChannels > 0)
+                    deviceInfos.push_back(pdi);
+            } else {
+                if (pdi->maxInputChannels > 0)
+                    deviceInfos.push_back(pdi);
+            }
+        } else
+            deviceInfos.push_back(pdi);
+    }
+
+    int numSelectedDevices = deviceInfos.size();
```

Comment:
personally, i would just use `deviceInfos.size()` instead of creating another local variable. it's the same number of characters, and typically having fewer local variables (within reason) makes code easier to understand. when you see `numSelectedDevices` below, it's not immediately obvious that it has a relationship to `deviceInfos`, and it's definitely not obvious that that relationship stays intact. small point, but just wanted to state my preference